### PR TITLE
Add Supervised Fine Tuning Train Operator, Hook, Tests, Docs

### PR DESCRIPTION
**Pull request by CYarros10**
This pull request adds the following:

**SupervisedFineTuningHook**: Hook for Google Cloud Vertex AI Supervised Fine Tuning APIs.
**SupervisedFineTuningTrainOperator**: Use the Google Cloud Supervised Fine Tuning API to create a tuning job.

About  [Model tuning](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/tuning): a crucial process in adapting Gemini to perform specific tasks with greater precision and accuracy. Model tuning works by providing a model with a training dataset that contains a set of examples of specific downstream tasks.

A sample DAG containing these operators could look like:
JSONL training data arrives in GCS >> GCSObjectExistenceSensor >> SupervisedFineTuningTrainOperator >> GenerativeModelGenerateContentOperator

### CODE REVIEW

**Comment by CYarros10** [comment]
Conflicting tests:

[Static checks, mypy, docs / Static checks](https://github.com/apache/airflow/actions/runs/10603244920/job/29387419328?pr=41807#logs) : 

```
Run 'ruff' for extremely fast Python linting.......................................Failed
- hook id: ruff
- exit code: 1

All checks passed!
tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py:27:1: E402 Module level import not at top of file
   |
25 |   pytest.importorskip("google.cloud.aiplatform_v1")
26 |   
27 | / from airflow.providers.google.cloud.hooks.vertex_ai.supervised_fine_tuning import (
28 | |     SupervisedFineTuningHook,
29 | | )
   | |_^ E402
30 |   from tests.providers.google.cloud.utils.base_gcp_mock import (
31 |       mock_base_gcp_hook_default_project_id,
```

[Special tests / Pydantic removed test / All:Pydantic-Removed-Postgres:12:3.8: API Always BranchExternalPython BranchPythonVenv CLI Core ExternalPython Operators Other PlainAsserts](https://github.com/apache/airflow/actions/runs/10602210989/job/29384161799#logs) : 

```
_ ERROR collecting tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py _
ImportError while importing test module '/opt/airflow/tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py:24: in <module>
    from airflow.providers.google.cloud.hooks.vertex_ai.supervised_fine_tuning import (
airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py:25: in <module>
    import vertexai
E   ModuleNotFoundError: No module named 'vertexai'
```

File: tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from unittest import mock
+
+import pytest
+
+# For no Pydantic environment, we need to skip the tests
+pytest.importorskip("google.cloud.aiplatform_v1")
```
Comment (CYarros10) [inline tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py]: Conflicting tests when evaluating `airflow/tests/providers/google/cloud/operators/vertex_ai/test_supervised_fine_tuning.py` and `airflow/tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py` ...

[Static checks, mypy, docs / Static checks](https://github.com/apache/airflow/actions/runs/10603244920/job/29387419328?pr=41807#logs) : 

```
Run 'ruff' for extremely fast Python linting.......................................Failed
- hook id: ruff
- exit code: 1

All checks passed!
tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py:27:1: E402 Module level import not at top of file
   |
25 |   pytest.importorskip("google.cloud.aiplatform_v1")
26 |   
27 | / from airflow.providers.google.cloud.hooks.vertex_ai.supervised_fine_tuning import (
28 | |     SupervisedFineTuningHook,
29 | | )
   | |_^ E402
30 |   from tests.providers.google.cloud.utils.base_gcp_mock import (
31 |       mock_base_gcp_hook_default_project_id,
```

[Special tests / Pydantic removed test / All:Pydantic-Removed-Postgres:12:3.8: API Always BranchExternalPython BranchPythonVenv CLI Core ExternalPython Operators Other PlainAsserts](https://github.com/apache/airflow/actions/runs/10602210989/job/29384161799#logs) : 

```
_ ERROR collecting tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py _
ImportError while importing test module '/opt/airflow/tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/providers/google/cloud/hooks/vertex_ai/test_supervised_fine_tuning.py:24: in <module>
    from airflow.providers.google.cloud.hooks.vertex_ai.supervised_fine_tuning import (
airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py:25: in <module>
    import vertexai
E   ModuleNotFoundError: No module named 'vertexai'
```

File: airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+class SupervisedFineTuningHook(GoogleBaseHook):
+    """Hook for Google Cloud Vertex AI Supervised Fine Tuning APIs."""
+
+    def __init__(
+        self,
+        gcp_conn_id: str = "google_cloud_default",
+        impersonation_chain: str | Sequence[str] | None = None,
+        **kwargs,
+    ):
+        if kwargs.get("delegate_to") is not None:
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py]: I think this deprecation warning is not needed for the new Hook. Could you delete it?

File: airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        learning_rate_multiplier: float | None = None,
+        project_id: str = PROVIDE_PROJECT_ID,
+    ) -> SupervisedTuningJob:
+        """
+        Use the Supervised Fine Tuning API to create a tuning job.
+
+        :param source_model: Required. A pre-trained model optimized for performing natural
+            language tasks such as classification, summarization, extraction, content
+            creation, and ideation.
+        :param training_dataset: Required. Cloud Storage URI of your training dataset. The dataset
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py]:
```suggestion
        :param train_dataset: Required. Cloud Storage URI of your training dataset. The dataset
```

File: airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class SupervisedFineTuningTrainOperator(GoogleCloudBaseOperator):
+    """
+    Use the Supervised Fine Tuning API to create a tuning job.
+
+    :param source_model: Required. A pre-trained model optimized for performing natural
+        language tasks such as classification, summarization, extraction, content
+        creation, and ideation.
+    :param training_dataset: Required. Cloud Storage URI of your training dataset. The dataset
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py]:
```suggestion
    :param train_dataset: Required. Cloud Storage URI of your training dataset. The dataset
```

File: airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            adapter_size=self.adapter_size,
+            learning_rate_multiplier=self.learning_rate_multiplier,
+            tuned_model_display_name=self.tuned_model_display_name,
+        )
+
+        self.log.info("Tuned Model Name: %s", response.tuned_model_name)
+        self.log.info("Tuned Model Endpoint Name: %s", response.tuned_model_endpoint_name)
+
+        self.xcom_push(context, key="tuned_model_name", value=response.tuned_model_name)
+        self.xcom_push(context, key="tuned_model_endpoint_name", value=response.tuned_model_endpoint_name)
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py]: Could you add a return result? I think in your case it should be `SupervisedTuningJob` in dict format. For example how it looks for Pipeline Job: https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/vertex_ai/pipeline_job.py#L204C1-L206C1

**Comment by MaksYermak** [comment]
@CYarros10 What do you think about adding Links to this Operator and, maybe, for the previous operators related to generative AI? It is an example of code how it looks for PipelineJob https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/vertex_ai/pipeline_job.py#L115 and https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/links/vertex_ai.py#L329

File: airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py (hunk 2 above)
Comment (CYarros10) [reply airflow/providers/google/cloud/hooks/vertex_ai/supervised_fine_tuning.py]: Done!

File: airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py (hunk 5 above)
Comment (CYarros10) [reply airflow/providers/google/cloud/operators/vertex_ai/supervised_fine_tuning.py]: Thanks for the advice, added in latest commits

**Comment by CYarros10** [comment]
> @CYarros10 What do you think about adding Links to this Operator and, maybe, for the previous operators related to generative AI? It is an example of code how it looks for PipelineJob https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/vertex_ai/pipeline_job.py#L115 and https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/links/vertex_ai.py#L329

I think this is a great idea, I would love to do this but want to prioritize [CountTokensAPI](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/count-tokens?hl=en) and [EvaluationAPI](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/evaluation?hl=en) as these could be part of a broader LLM pipeline with SupervisedTuningTrainOperator and GenerativeModelGenerateContentOperator

**Comment by CYarros10** [comment]
Refactored the code in this PR to be included in `airflow/airflow/providers/google/cloud/operators/vertex_ai/generative_model.py` - to keep all generative AI / generative model operations in one place. as well as hooks, tests, docs, etc.

**Review by MaksYermak — APPROVED** [review]
LGTM

**Review by potiuk — APPROVED** [review]
Nice one

**Pull request merged**